cloudinit/sources/DataSourceGCE.py

```diff
@@ -88,13 +88,39 @@ def __init__(self, sys_cfg, distro, paths):
 
     def _get_data(self):
         url_params = self.get_url_params()
-        network_context = noop()
         if self.perform_dhcp_setup:
-            network_context = EphemeralDHCPv4(
-                self.distro,
-                self.fallback_interface,
+            candidate_nics = net.find_candidate_nics(
+                default_primary_interface=DEFAULT_PRIMARY_INTERFACE
             )
-        with network_context:
+            assert (
+                len(candidate_nics) >= 1
+            ), "The instance has to have at least one candidate NIC"
+            for candidate_nic in candidate_nics:
+                network_context = EphemeralDHCPv4(
+                    self.distro,
+                    iface=candidate_nic,
+                )
+                try:
+                    with network_context:
+                        ret = util.log_time(
+                            LOG.debug,
+                            "Crawl of GCE metadata service",
+                            read_md,
```

Comment:
If `read_md` fails for some reason do we want to catch the exception and continue to try other candidates? If we hit some error here it looks like we are on the uncaught exception path and avoid iterating through any remaining candidates.

Reply:
Catched and added debug log with the exception.